core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -759,8 +762,107 @@ class TransactionsTest extends IntegrationTestHarness {
   }
 
   @ParameterizedTest
-  @ValueSource(strings = Array("zk", "kraft"))
-  def testFailureToFenceEpoch(quorum: String): Unit = {
+  @CsvSource(Array("kraft, true"))
+  def testBumpTransactionalEpochWithTV2Enabled(quorum: String, isTV2Enabled: Boolean): Unit = {
+    val producer = createTransactionalProducer("transactionalProducer",
+      deliveryTimeoutMs = 5000, requestTimeoutMs = 5000)
+    val consumer = transactionalConsumers.head
+
+    try {
+      // Create a topic with RF=1 so that a single broker failure will render it unavailable
+      val testTopic = "test-topic"
+      createTopic(testTopic, numPartitions, 1, new Properties)
+      val partitionLeader = TestUtils.waitUntilLeaderIsKnown(brokers, new TopicPartition(testTopic, 0))
+
+      producer.initTransactions()
+
+      // First transaction: commit
+      producer.beginTransaction()
+      producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(testTopic, 0, "4", "4", willBeCommitted = true))
+      producer.commitTransaction()
+
+      // Get producerId and epoch after first commit
+      val log = brokers(partitionLeader).logManager.getLog(new TopicPartition(testTopic, 0)).get
+      val producerStateManager = log.producerStateManager
+      val activeProducersIter = producerStateManager.activeProducers.entrySet().iterator()
+      assertTrue(activeProducersIter.hasNext)
+      var producerStateEntry = activeProducersIter.next().getValue
+      val producerId = producerStateEntry.producerId
+      var previousProducerEpoch = producerStateEntry.producerEpoch
+
+      // Second transaction: abort
+      producer.beginTransaction()
+      producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(topic1, null, "2", "2", willBeCommitted = false))
+
+      killBroker(partitionLeader)
+      val failedFuture = producer.send(TestUtils.producerRecordWithExpectedTransactionStatus(testTopic, 0, "3", "3", willBeCommitted = false))
+      Thread.sleep(6000)
```

Comment:
nit: I know the other test does this, but we should try to avoid thread.sleep if possible. Maybe something like wait until the future returns.

Reply:
Test is super flaky when using waitUntilTrue for the future to complete